### PR TITLE
fix: IaC SARIF output for GitHub Action [CC-957]

### PR DESCRIPTION
**Pull request by teodora-sandu**
#### What does this PR do?
In https://github.com/snyk/snyk/pull/1988 we fixed our GitHub Action by making the GA-ed SARIF output consistent with the legacy one. But there were still bugs left behind from the legacy output This PR completely removes all bugs to do with the file paths in the SARIF output.

#### Where should the reviewer start?
The tests have been refactored to make them easier to read, but also modified after implementing the fixes. Each commit contains its respective change to the functionality and tests, so to understand why these changes were made go through each `fix` commit.

#### How should this be manually tested?
Build using `npm run build` and clone https://github.com/teodora-sandu/test. Change directory into `nested` and run the following commands one by one. The result of these commands can be found in the attached ZIP.
```
snyk iac test --sarif-file-output=output-ga.sarif
snyk iac test --sarif-file-output=output-legacy.sarif --legacy
snyk iac test --sarif-file-output=output-ga-path.sarif .
snyk iac test --sarif-file-output=output-legacy-path.sarif --legacy .

snyk-dev iac test --sarif-file-output=output-ga.sarif
snyk-dev iac test --sarif-file-output=output-legacy.sarif --legacy
snyk-dev iac test --sarif-file-output=output-ga-path.sarif .
snyk-dev iac test --sarif-file-output=output-legacy-path.sarif --legacy .
```
If you compare the `output-ga.sarif` before and after, you'll see that the after looks exactly like the `output-ga-path.sarif`, which is the correct one, used by the GitHub action.

#### Any background context you want to provide?
- PR where the now removed FIXME was added: https://github.com/snyk/snyk/pull/1550/files
- thread explaining that URI's should use `/` in Windows instead of `\`: https://stackoverflow.com/questions/1589930/so-what-is-the-right-direction-of-the-paths-slash-or-under-windows

#### What are the relevant tickets?
https://snyksec.atlassian.net/browse/CC-957

#### Screenshots
Before:
[before.zip](https://github.com/snyk/snyk/files/6702749/before.zip)
Screenshot 2021-06-23 at 18 15 39
Screenshot 2021-06-23 at 18 17 19
Screenshot 2021-06-23 at 18 15 57

After:
[after.zip](https://github.com/snyk/snyk/files/6702750/after.zip)
Screenshot 2021-06-23 at 18 15 16
Screenshot 2021-06-23 at 18 17 10
Screenshot 2021-06-23 at 18 16 10

### CODE REVIEW

File: src/cli/commands/test/iac-output.ts [hunk 1]
```diff
@@ -125,14 +126,14 @@ const PROJECT_ROOT_KEY = 'PROJECTROOT';
...
+    : pathLib.resolve('.');
   const issues = iacTestResponses.reduce((collect: ResponseIssues, res) => {
     if (res.result) {
-      // FIXME: For single file tests the targetFile includes the full file
-      // path, for directory tests only the filename is returned and we need
-      // too build the path manually.
-      const targetPath = res.targetFile.startsWith(res.path)
-        ? pathLib.join(res.targetFile)
-        : pathLib.join(res.path, res.targetFile);
+      const targetPath = res.targetFile.replace(/\\/g, '/');
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-output.ts:136]: `targetFile` is the relative path of the scanned file (regardless of whether it was provided as a relative or absolute path): https://github.com/snyk/snyk/blob/master/src/cli/commands/test/iac-local-execution/results-formatter.ts#L136.
https://github.com/snyk/snyk/blob/master/src/cli/commands/test/iac-local-execution/results-formatter.ts#L131

File: test/jest/acceptance/iac/file-output.spec.ts [hunk 2]
```diff
@@ -174,13 +164,18 @@ describe('iac test --sarif-file-output', () => {
...
-      const actualPhyisicalLocation =
+      const actualPhysicalLocation =
         jsonObj?.runs?.[0].results[0].locations[0].physicalLocation
           .artifactLocation.uri;
-      expect(actualProjectRoot).toEqual(expectedProjectRoot);
-      expect(actualPhyisicalLocation).toEqual(expectedPhysicalLocation);
+      const actualProjectRoot =
+        jsonObj?.runs?.[0].originalUriBaseIds.PROJECTROOT.uri;
+      expect(actualPhysicalLocation).toEqual(
+        path.join(expectedPhysicalLocation).replace(/\\/g, '/'),
```
Comment (teodora-sandu) [inline test/jest/acceptance/iac/file-output.spec.ts:173]: I've moved the duplicated `path.join` and `path.resolve` logic from the test setup to the test, so it's easier to read what the test takes

File: test/jest/acceptance/iac/file-output.spec.ts [hunk 3]
```diff
@@ -174,13 +164,16 @@ describe('iac test --sarif-file-output', () => {
...
-        jsonObj?.runs?.[0].originalUriBaseIds.PROJECTROOT.uri;
-      const actualPhyisicalLocation =
+      const actualPhysicalLocation =
         jsonObj?.runs?.[0].results[0].locations[0].physicalLocation
           .artifactLocation.uri;
-      expect(actualProjectRoot).toEqual(expectedProjectRoot);
-      expect(actualPhyisicalLocation).toEqual(expectedPhysicalLocation);
+      const actualProjectRoot =
+        jsonObj?.runs?.[0].originalUriBaseIds.PROJECTROOT.uri;
+      expect(actualPhysicalLocation).toEqual(expectedPhysicalLocation);
```
Comment (teodora-sandu) [inline test/jest/acceptance/iac/file-output.spec.ts:172]: I've moved the duplicated `path.join` and `path.resolve` logic from the test setup to the test, so it's easier to read what the test takes

File: test/jest/acceptance/iac/file-output.spec.ts [hunk 4]
```diff
@@ -174,13 +172,16 @@ describe('iac test --sarif-file-output', () => {
...
+      expect(actualProjectRoot).toEqual(
+        'file://' +
+          path.join(path.resolve(expectedProjectRoot), '/').replace(/\\/g, '/'),
+      );
```
Comment (teodora-sandu) [inline test/jest/acceptance/iac/file-output.spec.ts:184]: I've moved the duplicated `path.join` and `path.resolve` logic from the test setup to the test, so it's easier to read what the test takes

File: src/cli/commands/test/iac-output.ts (hunk 1 above)
Comment (aron) [reply src/cli/commands/test/iac-output.ts:136]: Great note, I think this comment is worth adding inline above the code

**Review by aron — APPROVED** [review]
👏 Very nice work! This is so much clearer.

**Pull request merged**